tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+package suite
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/batch/v1"
+	coordination "k8s.io/api/coordination/v1"
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	"github.com/nginxinc/nginx-gateway-fabric/tests/framework"
+)
+
+const (
+	teaURL             = "https://cafe.example.com/tea"
+	coffeeURL          = "http://cafe.example.com/coffee"
+	nginxContainerName = "nginx"
+	ngfContainerName   = "nginx-gateway"
+)
+
+var _ = Describe("Graceful Recovery test", Ordered, Label("nfr", "graceful-recovery"), func() {
+	files := []string{
+		"graceful-recovery/cafe.yaml",
+		"graceful-recovery/cafe-secret.yaml",
+		"graceful-recovery/gateway.yaml",
+		"graceful-recovery/cafe-routes.yaml",
+	}
+	ns := &core.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "graceful-recovery",
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+	})
+
+	AfterAll(func() {
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+	})
+
+	It("recovers when NGF container is restarted", func() {
+		runRecoveryTest(ngfContainerName, files, ns)
+	})
+
+	It("recovers when nginx container is restarted", func() {
+		runRecoveryTest(nginxContainerName, files, ns)
+	})
+})
+
+func runRecoveryTest(containerName string, files []string, ns *core.Namespace) {
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	leaseName, err := getLeaderElectionLeaseHolderName()
```

Comment:
I'd add a comment here why we confident that the NGF has the lease here
why not can election happen 5 seconds after this?

Reply:
Does this sound about right?

"Since we have already deployed resources and ran `resourceManager.WaitForAppsToBeReady(ns.Name)` earlier, we know that the applications are ready at this point. This could only be the case if NGF has written statuses, which could only be the case if NGF has the leader lease. Since there is only one instance of NGF in this test, we can be certain that this is the correct lease holder name"